compiler/src/dotty/tools/dotc/transform/localopt/Valify.scala

```diff
@@ -7,87 +7,193 @@ import core.Types._
 import core.Flags._
 import ast.Trees._
 import scala.collection.mutable
+import java.util.IdentityHashMap
+import core.NameKinds.LocalOptValifyName
 
-/** Rewrite vars with exactly one assignment as vals.
+/** Rewrite vars as vals when possible.
  *
- *  @author DarkDimius, OlivierBlanvillain
+ *  @author gan74
  */
 class Valify(val simplifyPhase: Simplify) extends Optimisation {
   import ast.tpd._
 
-  // Either a duplicate or a read through series of immutable fields.
-  val defined: MutableSymbolMap[ValDef] = newMutableSymbolMap
+  object VarDef {
+    def unapply(t: ValDef)(implicit ctx: Context): Option[(Symbol, Tree)] =
+      if (isVar(t.symbol)) Some(t.symbol, t.rhs)
+      else None
+  }
 
-  val firstRead: MutableSymbolMap[RefTree] = newMutableSymbolMap
+  object VarAssign {
+    def unapply(t: Assign)(implicit ctx: Context): Option[(Symbol, Tree)] =
+      if (t.lhs.isInstanceOf[Ident] && isVar(t.lhs.symbol)) Some(t.lhs.symbol, t.rhs)
+      else None
+  }
 
-  val firstWrite: MutableSymbolMap[Assign] = newMutableSymbolMap
 
-  val secondWrite: MutableSymbolMap[Assign] = newMutableSymbolMap
+  def clear(): Unit = ()
 
-  def clear(): Unit = {
-    defined.clear()
-    firstRead.clear()
-    firstWrite.clear()
-    secondWrite.clear()
-  }
+  def visitor(implicit ctx: Context): Tree => Unit = NoVisitor
 
-  def visitor(implicit ctx: Context): Tree => Unit = {
-    case t: ValDef if t.symbol.is(Mutable, Lazy) && !t.symbol.is(Method) && !t.symbol.owner.isClass =>
-      if (isPureExpr(t.rhs))
-        defined(t.symbol) = t
-
-    case t: RefTree if t.symbol.exists && !t.symbol.is(Method) && !t.symbol.owner.isClass =>
-      if (!firstWrite.contains(t.symbol)) firstRead(t.symbol) = t
-
-    case t @ Assign(l, expr) if !l.symbol.is(Method) && !l.symbol.owner.isClass =>
-      if (!firstRead.contains(l.symbol)) {
-        if (firstWrite.contains(l.symbol)) {
-          if (!secondWrite.contains(l.symbol))
-            secondWrite(l.symbol) = t
-        } else if (!expr.existsSubTree(x => x match {
-          case tree: RefTree if x.symbol == l.symbol => firstRead(l.symbol) = tree; true
-          case _ => false
-        })) {
-          firstWrite(l.symbol) = t
+
+  def transformer(implicit ctx: Context): Tree => Tree = {
+    case blk @ Block(stats, expr) => 
+
+      val valified = mutable.Map[Symbol, Symbol]()
+
+      // did we replace anything ?
+      var replaced = false
+
+      // recursively replace vars in tree
+      def transform(tree: Tree) = {
+        if (valified.isEmpty) tree
+        else {
+          // abort the valification for vars that are modified in a nested (ie: potentially conditional) subtree
+          tree.foreachSubTree { 
+            case VarAssign(sym, _) if valified.contains(sym) => 
+              valified -= sym
+            case _ => 
+          }
+
+          if (valified.isEmpty) tree
+          else new TreeMap() {
```

Comment:
Same comment that on #3511, this quadratic traversal needs to be addressed before merge